packages/slate/src/models/text.js

```diff
@@ -186,6 +186,16 @@ class Text extends Record(DEFAULTS) {
           const offset = o
           o += length
 
+          // If `type` is `placeholder`, we can be sure that there
+          // are no other leaves and it's safe to bail out early.
+          // This step is necessary (for now), otherwise placeholder
+          // decoration won't be added to leaf and doesn't render.
+          if (format.type === 'placeholder') {
+            leaf[kind].push(format)
+            next.push(leaf)
+            continue
+          }
```

Comment:
@ianstormtaylor I totally agree and I'm not 100% sure yet how to get around this. The issue is that `renderMark` was not invoked, so I had to change it to decoration. But the issue with that is there's no text content in leaf and if there's no text, decoration isn't getting added and ends up not rendering.

Reply:
@adjourn gotcha. In that case, we need to fix the decoration not getting added and rendering. It sounds like maybe the root cause is decorations with zero lengths, and that's what should be solved instead. And then we can remove this bit with the hardcoded `'placeholder'` logic.

Reply:
@ianstormtaylor I guess it makes sense in some situations that decorations and annotation work with zero width content (both are added based on same logic). 

Alternatively I could look into why `renderMark` won't invoke. What's the correct returned object structure in `decorateNode` to add mark (instead of decoration) after the annotation update?

Reply:
@adjourn well `renderMark` is no longer used for this use case. The `decorateNode` handler returns `Decoration` objects (or their JSON equivalents) and those objects are now rendered with the `renderDecoration` handler. So marks should be completely separate from this logic now.

Actually I think the newest commit is almost there. Re-reading the existing code, I *think* the logic should actually be slightly further down. Right now we have:

```js
// If the range starts after the leaf, or ends before it, continue.
if (start.offset > offset + length || end.offset <= offset) {
  next.push(leaf)
  continue
}

// If the range encompases the entire leaf, add the format.
if (start.offset <= offset && end.offset >= offset + length) {
  leaf[kind].push(format)
  next.push(leaf)
  continue
}
```

But the issue is that the first `end.offset <= offset` is getting hit, resulting in the leaf being skipped. I think that first condition needs a small tweak to ensure that it misses in this case? Because it looks like the second condition would have correctly caught this case if the first hadn't exited early.

Actually... I think maybe those two conditions could be flipped, so that the second one runs first? And that might solve it without any other changes.

Reply:
@ianstormtaylor can't believe I didn't catch this. See #2819 (this one got super messy).